### PR TITLE
Add config evaluation events to OTel spans

### PR DESCRIPTION
**Pull request by passichenko**
# Motivation
It will enable users to query traces based on config values that were evaluated when processing a request.
More about background and motivation: https://github.com/open-telemetry/semantic-conventions/blob/main/docs/feature-flags/feature-flags-spans.md
see https://github.com/lekkodev/core/pull/478 for example usage

# Change
I didn't follow conventions exactly, but we can change that later. Basically if OTel tracing is enabled, SDK will try attach config evaluation event to the active span. If there is no active span it will be a no-op.
Event attributes:
- `config.key` is config name
- `config.string_value` is string representation of config's value, for now it's only logged for bool, int and float types
- `config.version` is sha of the last commit that modified the config, it can be used to lookup exact version of the config that was used

## TODO
- add more metadata, like value path (what override was used)

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -43,51 +46,86 @@ type CloseFunc func(context.Context) error
...
+func newClient(provider Provider, options ClientOptions) (*client, CloseFunc) {
+	return &client{
+		provider:          provider,
+		startupContext:    options.StartupContext,
+		enableOTelTracing: options.EnableOTelTracing,
+	}, provider.Close
+}
+
+func (c *client) addTracingEvent(ctx context.Context, key, stringValue string) {
+	if !c.enableOTelTracing {
```
[the remark this answers is not in the document]
Comment (passichenko) [reply client/client.go:79]: I looked around it looks like the convention is that OTel integration should be opt-in.

File: client/client.go [hunk 2]
```diff
@@ -43,51 +46,90 @@ type CloseFunc func(context.Context) error
...
+	}
+	span := trace.SpanFromContext(ctx)
+	attributes := []attribute.KeyValue{
+		attribute.String("config.key", key),
+	}
+	if len(stringValue) > 0 {
+		attributes = append(attributes, attribute.String("config.string_value", stringValue))
+	}
+	if len(version) > 0 {
+		attributes = append(attributes, attribute.String("config.version", version))
```
Comment (passichenko) [inline client/client.go:90]: I started with `version` here, but then I named it `LastUpdateCommitSHA` in our protos. I'm not sure which one is better, any suggestions?

File: client/client.go [hunk 3]
```diff
@@ -43,51 +46,86 @@ type CloseFunc func(context.Context) error
...
+	// todo: use commit/config hash it value is too big
+	c.addTracingEvent(ctx, key, res)
+	return res, err
 }
 
 func (c *client) GetProto(ctx context.Context, namespace, key string, result proto.Message) error {
-	return c.provider.GetProto(c.wrap(ctx), key, namespace, result)
+	err := c.provider.GetProto(c.wrap(ctx), key, namespace, result)
+	// todo: use commit/config hash
+	c.addTracingEvent(ctx, key, "<proto>")
```
[the remark this answers is not in the document]
Comment (passichenko) [reply client/client.go]: I changed it to log value for bool, int and float types only. I also added `version` (which is sha of the last update), see updated description.

File: client/client.go (hunk 2 above)
Comment (shubhitms) [reply client/client.go:90]: i think the latter is more descriptive and works well

File: client/provider.go [hunk 4]
```diff
@@ -27,14 +27,18 @@ import (
...
+
 // A provider evaluates configuration from a number of sources.
 type Provider interface {
-	GetBool(ctx context.Context, key string, namespace string) (bool, error)
-	GetInt(ctx context.Context, key string, namespace string) (int64, error)
-	GetFloat(ctx context.Context, key string, namespace string) (float64, error)
-	GetString(ctx context.Context, key string, namespace string) (string, error)
-	GetProto(ctx context.Context, key string, namespace string, result proto.Message) error
-	GetJSON(ctx context.Context, key string, namespace string, result interface{}) error
+	GetBool(ctx context.Context, key string, namespace string) (bool, Metadata, error)
```
[the remark this answers is not in the document]
Comment (DavidSGK) [reply client/provider.go:36]: ^ +1 on this, I think keeping a unified options pattern is important to reduce confusion.

[the remark this answers is not in the document]
Comment (passichenko) [reply client/provider.go:36]: I considered that initially. The downside is that we'd need to implement tracing per-provider and also make sure that it's working and won't break with some other changes later.

[the remark this answers is not in the document]
Comment (shubhitms) [reply client/provider.go:36]: we can minimize the duplication by writing a single trace emitter and injecting it into every provider. it's a slight cost we pay for keeping all the options in one place

[the remark this answers is not in the document]
Comment (passichenko) [reply client/provider.go:36]: I started on this, looks a bit cleaner. But at the same time I started wondering why do we need client as a separate abstraction over provider? As of now, the only thing that client does is to handle startup context.

**Comment by passichenko** [comment]
abandoned in favor of simplified version in https://github.com/lekkodev/go-sdk/pull/119
will probably do some of the additional things like logging values to events in separate PRs

**Pull request closed without merging**